src/services/AsyncStorage.ts

```diff
@@ -56,6 +59,34 @@ export const removeCustomDiscipline = async (customDiscipline: string): Promise<
   await setCustomDisciplines(disciplines)
 }
 
+export const getExerciseProgress = async (): Promise<Array<ExerciseProgress>> => {
+  const progress = await AsyncStorage.getItem(exerciseProgress)
+  return progress ? JSON.parse(progress) : []
+}
+
+export const setExerciseProgress = async (progress: ExerciseProgress): Promise<void> => {
+  const savedProgress = await getExerciseProgress()
+  const index = savedProgress.findIndex(
```

Comment:
Can't we save a map `id -> ExerciseProgress` instead of an `ExerciseProgress[]` array? I think that would make mutation a lot easier. This logic seems to be quite complicated.

Reply:
Yes we can use a Map, but it is really ugly to save / read Maps from JSON. This would be the working code for Maps:
`
const replacer = (key: any, value: any) => {
  if (value instanceof Map) {
    return {
      dataType: 'Map',
      value: Array.from(value.entries()) // or with spread: value: [...value]
    }
  }
  return value
}

const reviver = (key: any, value: any) => {
  if (typeof value === 'object' && value !== null) {
    if (value.dataType === 'Map') {
      return new Map(value.value)
    }
  }
  return value
}

export const getExerciseProgress = async (): Promise<Progress> => {
  const progress = await AsyncStorage.getItem(progressKey)
  return progress ? JSON.parse(progress, reviver) : new Map()
}

const setExerciseProgress = async (disciplineId: number, exerciseKey: ExerciseKey, score: number): Promise<void> => {
  const savedProgress = await getExerciseProgress()
  const disciplineProgress = savedProgress.get(disciplineId)
  if (disciplineProgress) {
    const exerciseProgress = disciplineProgress.get(exerciseKey)
    if (!exerciseProgress || exerciseProgress <= score) {
      disciplineProgress.set(exerciseKey, score)
      savedProgress.set(disciplineId, disciplineProgress)
    }
  } else {
    const exerciseProgress: ExerciseProgress = new Map()
    exerciseProgress.set(exerciseKey, score)
    savedProgress.set(disciplineId, exerciseProgress)
  }
  await AsyncStorage.setItem(progressKey, JSON.stringify(savedProgress, replacer))
`